File: src/main/java/game/Ladrao.java
package game;

public class Ladrao extends Peca {
    public Ladrao(String cor) {
        super(cor);
        this.nomePeca = "Ladrao";
    }

    @Override
    public boolean isMovimentoValido(Tabuleiro tabuleiro, int linhaInicial, int colunaInicial, int linhaFinal, int colunaFinal) {
        Peca pecaDestino = tabuleiro.getPeca(linhaFinal, colunaFinal);

        // Checagem do Movimento normal -Diagonal ATÉ 2 casa
        int diffLinha = Math.abs(linhaFinal - linhaInicial);
        int diffColuna = Math.abs(colunaFinal - colunaInicial);

        if (diffLinha != diffColuna) {
            return false;
        }

        if (diffLinha > 0 && diffLinha <= 2) {
            // Se o movimento for de 2 casas, o caminho precisa estar livre.
            if (diffLinha == 2) {
                int passoLinha = (linhaFinal > linhaInicial) ? 1 : -1;
                int passoColuna = (colunaFinal > colunaInicial) ? 1 : -1;

                // Verifica a casa intermediária
                if (tabuleiro.getPeca(linhaInicial + passoLinha, colunaInicial + passoColuna) != null) {
                    return false; // Caminho bloqueado, movimento inválido.
                }
            }

            if (pecaDestino == null || !pecaDestino.getCor().equals(this.getCor())) {
                return true;
            }
        }

        return false; 
    }

    @Override
    public String toString() {
        return "L{}";
    }
}

File: Projeto JS/src/game/Ladrao.js
import { Peca } from './Peca.js';


export class Ladrao extends Peca {
    constructor(cor) {
        super(cor);
        this.nomePeca = "Ladrao";
    }

    isMovimentoValido(tabuleiro, linhaInicial, colunaInicial, linhaFinal, colunaFinal) {
        const pecaDestino = tabuleiro.getPeca(linhaFinal, colunaFinal);

   
        const diffLinha = Math.abs(linhaFinal - linhaInicial);
        const diffColuna = Math.abs(colunaFinal - colunaInicial);


        if (diffLinha !== diffColuna) {
            return false;
        }

    
        if (diffLinha > 0 && diffLinha <= 2) {

            if (diffLinha === 2) {
                const passoLinha = (linhaFinal > linhaInicial) ? 1 : -1;
                const passoColuna = (colunaFinal > colunaInicial) ? 1 : -1;

               
                if (tabuleiro.getPeca(linhaInicial + passoLinha, colunaInicial + passoColuna) !== null) {
                    return false; 
                }
            }

           
            if (pecaDestino === null || pecaDestino.getCor() !== this.getCor()) {
                return true;
            }
        }

       
        return false;
    }

    toString() {
        return "L{}";
    }
}